src/g/selection/SelectionCol.js

```diff
@@ -200,34 +225,328 @@ const SelectionManager = Collection.extend({
+      const { type: selType } = model.attributes;
+      switch (selType) {
+        case 'row': {
+          const { seqId } = model.attributes;
+          selectionData.selectedRows.add(seqId);
+          break;
+        }
+        case 'column': {
+          const { xStart, xEnd } = model.attributes;
+          for (let j = xStart; j <= xEnd; j++) {
+            selectionData.selectedColumns.add(j);
+          }
+          break;
+        }
+        case 'pos': {
+          const { xStart, seqId, xEnd } = model.attributes;
+          for (let j = xStart; j <= xEnd; j++) {
+            if (!selectedPositionsSet.has(`${seqId}-${j}`)) {
+              selectionData.selectedPositions.push({ seqId: seqId, colIdx: j });
+              selectedPositionsSet.add(`${seqId}-${j}`);
+            }
+          }
+          break;
+        }
+        case 'label': {
+          const { seqId } = model.attributes;
+          selectionData.selectedLabels.add(seqId);
+          break;
         }
+        default:
+          break;
+      }
+    })
+    return {
+      selectedRows: Array.from(selectionData.selectedRows),
+      selectedColumns: Array.from(selectionData.selectedColumns),
+      selectedPositions: selectionData.selectedPositions,
+      selectedLabels: Array.from(selectionData.selectedLabels),
+    };
+  },
+
+  setSelectionData: function(selectionData, silent = false) {
+    const models = [];
+    selectionData.selectedRows.forEach((seqId) => {
+      models.push(new rowsel({seqId}));
+    });
+
+    selectionData.selectedColumns.forEach((xStart) => {
+      models.push(new columnsel({xStart, xEnd: xStart}));
+    });
+
+    selectionData.selectedPositions.forEach(({ seqId, colIdx }) => {
+      models.push(new possel({xStart: colIdx, xEnd: colIdx, seqId}));
+    });
+
+    selectionData.selectedLabels.forEach((seqId) => {
+      models.push(new labelsel({seqId}));
+    });
+
+    this.reset(models, {silent});
+  },
+
+  _refineSelections: function() {
+    // 1. Refine selections to remove any overlapping or duplicate selections.
+    // 2. Reduce contiguous column or position selections to groups of individual column or position selections.
+    const selectionData = this.getSelectionData();
+    this.setSelectionData(selectionData);
+  },
+
+  _getSelsWithLabelsForRows: function(selectionArr) {
+    let updatedSelectionArr = [];
+
+    selectionArr.forEach((sel) => {
+      if (sel.get("type") === "row") {
+        updatedSelectionArr.push(sel);
+        updatedSelectionArr.push(new labelsel({seqId: sel.get("seqId")}));
       } else {
-        if (selectedRowSeqIds.includes(selectionId)) {
-          const modelToRemove = this.models.find(m => m.get("seqId") === selectionId)
-          this.remove([modelToRemove])
-        } else {
-          this.reset([selection]);
+        updatedSelectionArr.push(sel);
+      }
+    });
+    return updatedSelectionArr;
+  },
+
+  _isSelectionValid: function(selection) {
```

Comment:
This function is used to invalidate `lastSelection` if it is out-of-bounds for the current set of sequences. I did this because we don't directly update `lastSelection` when we change selection data in MSA. So, `lastSelection` may not be valid for the new data.

Reply:
I see. Please also add this as a code comment.